slo_generator/backends/datadog.py

```diff
@@ -0,0 +1,171 @@
+        start = timestamp - window
+        end = timestamp
+        query = measurement['query']
+        query = self._fmt_query(query, window)
+        response = self.client.Metric.query(start=start, end=end, query=query)
+        LOGGER.debug(f"Result valid: {pprint.pformat(response)}")
+        sli_value = DatadogBackend.count(response)
+        return sli_value
+
+    def query_slo(self, timestamp, window, slo_config):
+        """Query SLO value from a given Datadog SLO.
+
+        Args:
+            timestamp (int): UNIX timestamp.
+            window (int): Window (in seconds).
+            slo_config (dict): SLO configuration.
+
+        Returns:
+            tuple: Good event count, bad event count.
+        """
+        slo_id = slo_config['backend']['measurement']['slo_id']
+        from_ts = timestamp - window
+        slo_data = self.client.ServiceLevelObjective.get(id=slo_id)
+        LOGGER.debug(
+            f"SLO data: {slo_id} | Result: {pprint.pformat(slo_data)}")
+        data = self.client.ServiceLevelObjective.history(id=slo_id,
+                                                         from_ts=from_ts,
+                                                         to_ts=timestamp)
+        LOGGER.debug(
+            f"Timeseries data: {slo_id} | Result: {pprint.pformat(data)}")
+        good_event_count = data['data']['series']['numerator']['sum']
+        valid_event_count = data['data']['series']['denominator']['sum']
+        bad_event_count = valid_event_count - good_event_count
+        return (good_event_count, bad_event_count)
+
+    @staticmethod
+    def _fmt_query(query, window, operator=None):
+        """Format Datadog query:
+
+        * If the Datadog expression has a `[window]` placeholder, replace it by
+        the current window. Otherwise, append it to the expression.
+
+        * If prefix / suffix operators are defined, apply them to the metric.
+
+        * If labels are defined, append them to existing labels.
+
+        Args:
+            query (str): Original query in YAML config.
+            window (int): Query window (in seconds).
+            operator (str): Operator (e.g: sum, avg, median, ...)
+
+        Returns:
+            str: Formatted query.
+        """
+        query = query.strip()
+        if operator:
+            query = f'{operator}:{query}'
+        if '[window]' in query:
+            query = query.replace('[window]', f'{window}')
+        LOGGER.debug(f'Query: {query}')
+        return query
+
+    @staticmethod
+    def count(timeseries):
+        """Count events in time series. If multiple values are returned, take
+        the average of all values.
+
+        Args:
+            :dict: Timeseries response from Datadog Metrics API endpoint.
+
+        Returns:
+            int: Event count.
+        """
+        try:
+            pointlist = timeseries['series'][0]['pointlist']
+            values = [point[1] for point in pointlist for i in point]
```

Comment:
Try and avoid multiple for clauses in the same list comprehension, as it makes it harder to read:

https://google.github.io/styleguide/pyguide.html#27-comprehensions--generator-expressions

Reply:
Refactor using a standard `for` loop